core/drivers/stm32_uart.c

```diff
@@ -45,17 +49,26 @@ static vaddr_t loc_chip_to_base(struct serial_chip *chip)
 static void loc_flush(struct serial_chip *chip)
 {
 	vaddr_t base = loc_chip_to_base(chip);
+	uint64_t timeout_ref = utimeout_init(FLUSH_TIMEOUT_US);
 
 	while (!(read32(base + UART_REG_ISR) & USART_ISR_TXFE))
-		;
+		if (utimeout_elapsed(FLUSH_TIMEOUT_US, timeout_ref))
+			return;
 }
 
 static void loc_putc(struct serial_chip *chip, int ch)
 {
 	vaddr_t base = loc_chip_to_base(chip);
+	uint64_t timeout_ref = utimeout_init(PUTC_TIMEOUT_US);
 
-	while (!(read32(base + UART_REG_ISR) & USART_ISR_TXE_TXFNF))
-		;
+	while (!(read32(base + UART_REG_ISR) & USART_ISR_TXE_TXFNF)) {
+		if (utimeout_elapsed(PUTC_TIMEOUT_US, timeout_ref)) {
```

Comment:
> This can be tricky: the peripheral clock could be OFF, the external bus pads could be mux-ed out.

OK. That argument is convincing enough :)

> Moreover the console bus may be controlled by the non-secure world in which case secure world cannot safely enable it [...]

Sure. I was not suggesting secure world should force-enable the device.